### PR TITLE
Fix remaining issues for simplified team plans

### PR DESCRIPTION
**Pull request by jankeromnes**
## Description

Improve and ship the simplified Team Billing. ✨

TODO:

- [x] Un-hide the new Team Billing menu entry
- [x] Polish the UI
- [x] Fix race condition on multiple team joins 🐞

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/10057

## How to test

1. Create a Team
2. Navigate to Team Settings > Team Billing -- it should look nice ✨
3. Upgrade your Team to either Professional or Unleashed
    - (provide bogus billing details, for example type "a" into every required field, and use the proposed fake credit card)
4. Invite two more members, who should join at almost exactly the same time (within ~30s)
5. Check Chargebee-test for new charges. There should be:
    - one "Pro-rated upgrade from 1 to 2"
    - one "Pro-rated upgrade from 2 to 3"

(If the bug isn't fixed, you'd see one "from 1 to 2" invoice, and another "from 1 to 3" invoice(!) that's double the amount)

Screenshot of a successful test:

Screenshot 2022-05-27 at 10 21 45

## Release Notes

```release-note
Fix, polish, and release new simplified Team Billing
```

## Documentation

- [x] /werft with-payment=true

### CODE REVIEW

**Comment by roboquat** [comment]
@jankeromnes: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/dashboard/src/teams/TeamBilling.tsx
Comment (gtsiolis) [inline components/dashboard/src/teams/TeamBilling.tsx:47]: **suggestion**: Probably not the best line I could comment about this, but what do you think of making the [section title](https://github.com/gitpod-io/gitpod/pull/10182/files#diff-fab4237d49423136bbbf60168ab7b05684291690d3b0679ae4fa02d51c858f55R157) more actionable and relatable to the action below?

From _No billing plan_ to _Upgrade team plan_. 💡

Comment (gtsiolis) [inline components/dashboard/src/teams/TeamBilling.tsx:47]: **suggestion**: Another irrelevant-line-to-comment-on: Could we rename team plan names and drop the duplicate and unnecessary _Team_ prefix, following the specs?

From _Team Professional_ to  _Professional_, etc.

Comment (gtsiolis) [inline components/dashboard/src/teams/TeamBilling.tsx:47]: **suggestion**: Another irrelevant-line-to-comment-on: Could we adjust font color and weight following the specs?
- I find it a bit hard to parse the information in the current state. I could not find the line of code to comment directly with the class names needed. Small changes can make a difference.
- Also, what do you think of moving the sum above the plan features as in the most recent design spec, to avoid implying that the label in the bottom right corner is an action button?

| BEFORE | AFTER |
|-|-|
| cards-before | cards-after |

Comment (gtsiolis) [inline components/dashboard/src/teams/TeamBilling.tsx:47]: **question**: Thoughts on rephrasing the subtitle from:

> Select a new billing plan for this team.

To:

> Upgrade team plan to access unlimited workspace hours, and more.

Comment (gtsiolis) [inline components/dashboard/src/teams/TeamBilling.tsx:47]: **question**: Last comment on irrelevant line of code: Instead of the hover state as used with the editor and theme in preferences, we could use here a clear action button for each case, but we'd need to drop the hover state on the card. WDYT?

| BEFORE | AFTER |
|-|-|
| Frame 1088 | Frame 1089 |

**Review by gtsiolis — COMMENTED** [review]
Thanks, @jankeromnes! Overall looks great. ✨ 

I couldn't find any code changes in this PR related to the UX comments left below, which made it harder to gather all these comments and find a place to add them. I assume these visual changes were part of another pull request, right?

Left some first round review comments. Let me know what you think.

**Comment by jankeromnes** [comment]
Wow, many thanks @gtsiolis for the in-depth UX review and spot-on suggestions! 🪄

> I couldn't find any code changes in this PR related to the UX comments left below, which made it harder to gather all these comments and find a place to add them. I assume these visual changes were part of another pull request, right?

Ah, whoops, indeed. Most of these rough visual changes were actually implemented in https://github.com/gitpod-io/gitpod/pull/8041, but were left hidden from users by removing the new menu entry. 🕶️

Thank you for still managing to do the in-depth review! 🙏

File: components/dashboard/src/teams/TeamBilling.tsx
Comment (jankeromnes) [reply components/dashboard/src/teams/TeamBilling.tsx:47]: I can easily hide the word "Team" in this page, if that's nicer visually, but we'll probably want to keep the word "Team" in all official team plans (e.g. to more easily differentiate them from individual plans in Chargebee, the DB, the Admin dashboard, etc)

Comment (gtsiolis) [reply components/dashboard/src/teams/TeamBilling.tsx:47]: Any option could suffice for this iteration. I'll leave this to you to decide, @jankeromnes!

Comment (jankeromnes) [reply components/dashboard/src/teams/TeamBilling.tsx]: I tried it, but then felt it looked too confusing -- using the same name as individual plans doesn't feel right. 🙈

So back to "Team" everywhere for now. Can be adjusted later, but also we'll deprecate these in favor of Usage-Based Billing in the not-too-distant future. 🔮

Comment (gtsiolis) [reply components/dashboard/src/teams/TeamBilling.tsx]: Thanks for giving this a try, @jankeromnes! Going with plan :a: with the prefix sounds good.

**Comment by jankeromnes** [comment]
Random lint config fail? 🤔 Let's try this again:

/werft run with-clean-slate-deployment

File: components/dashboard/src/teams/TeamBilling.tsx
Comment (gtsiolis) [inline components/dashboard/src/teams/TeamBilling.tsx:264]: **nitpick**: Minor spacing fix.

```suggestion
                                    <div className="mt-4 font-medium text-base text-gray-400 dark:text-gray-600">
```

File: components/dashboard/src/teams/TeamBilling.tsx
Comment (gtsiolis) [inline components/dashboard/src/teams/TeamBilling.tsx:201]: **praise**: Don't know what you changed here but thanks for fixing the pill element here to hug contents instead of filling the container width.

File: components/dashboard/src/components/DropDown.tsx
Comment (gtsiolis) [inline components/dashboard/src/components/DropDown.tsx:44]: **praise**: Thanks for changing this, @jankeromnes!

Comment (gtsiolis) [inline components/dashboard/src/components/DropDown.tsx:44]: **thought**: Wondering if we need the custom classes altogether here. 💭

File: components/dashboard/src/teams/TeamBilling.tsx
Comment (gtsiolis) [inline components/dashboard/src/teams/TeamBilling.tsx:256]: **thought**: Hoping in the future we can drop these by introducing component variants that include card size, etc. instead of hard-coding sizes like this. We'll get there!

File: components/dashboard/src/components/DropDown.tsx
Comment (gtsiolis) [inline components/dashboard/src/components/DropDown.tsx:49]: **thought**: This is interesting! Is this only about the color change? If yes, it could be fine to leave this without the color change to match the rest of the link dropdowns but it's fine to also leave this as is for for now. Your call.

| BEFORE | AFTER |
|-|-|
| link-before | link-after |

**Review by gtsiolis — APPROVED** [review]
UX LGTM, @jankeromnes! Left one last round of non-blocking comments.

**Comment by gtsiolis** [comment]
> /hold

@jankeromnes Also, regarding the hold comment in https://github.com/gitpod-io/gitpod/pull/10182#issuecomment-1141122366, in case it helps, this still needs a review from `@gitpod-io/engineering-webapp` to be merged.

> At least 1 approving review is required to merge this pull request.

File: components/dashboard/src/teams/TeamBilling.tsx
Comment (jankeromnes) [reply components/dashboard/src/teams/TeamBilling.tsx:256]: There are ways to achieve this with a single component, e.g. by using:
- `min-height` on the card (to allow it to grow arbitrary large depending on its content)
- `items-stretch` on the card container (to make all the cards the same height, even if one card has a lot of content but the others don't)

File: components/dashboard/src/components/DropDown.tsx
Comment (jankeromnes) [reply components/dashboard/src/components/DropDown.tsx:44]: I believe we do, because the context-menu sometimes need to be left-aligned or right-aligned, among other things.

File: components/dashboard/src/teams/TeamBilling.tsx
Comment (jankeromnes) [reply components/dashboard/src/teams/TeamBilling.tsx:201]: I removed a hard-coded `ml-2` from the component. 😇

File: components/dashboard/src/components/DropDown.tsx
Comment (jankeromnes) [reply components/dashboard/src/components/DropDown.tsx:49]: Yes, all these were only for the color change. 😅 But I think it's worth it, because with @jldec we found out that the grey currency switcher was very hard to notice when not link-color.

I think this is a nice improvement. ✨

**Comment by jankeromnes** [comment]
> At least 1 approving review is required to merge this pull request.

Oh. Sounds like a bug if it's true -- you should absolutely be able to approve at least dashboard fixes & improvements. 🚢

File: components/dashboard/src/components/DropDown.tsx
Comment (gtsiolis) [reply components/dashboard/src/components/DropDown.tsx:44]: It's better to  never use right-aligned items inside the dropdown menus. 💡 

Do you have a handy instance of a left-alignment you're referring to?

I've only seen the dropdown in the workspace start page when using a desktop editor that's using center alignment and this should also change, see https://github.com/gitpod-io/gitpod/issues/6910.

If this is about the alignment of the component itself with the rest of the elements around it, we could probably move the alignment control outside this component, right?

File: components/dashboard/src/components/DropDown.tsx
Comment (gtsiolis) [reply components/dashboard/src/components/DropDown.tsx:49]: DEAL. Let's keep it for now. Maybe we could update all link dropdowns to become blue in the long run. Or the other way around.

File: components/dashboard/src/teams/TeamBilling.tsx
Comment (jankeromnes) [reply components/dashboard/src/teams/TeamBilling.tsx:256]: Also, it's good to keep in mind that Tailwind is so nice because it allows directly "hard-coding" style in a nice way (as opposed to hiding style under multiple levels of abstractions, making them really hard to adjust afterwards).

Comment (jankeromnes) [reply components/dashboard/src/teams/TeamBilling.tsx:256]: > Long-term, to keep the component flexible I'd love to introduce a stepper property to increase the width or height without using Tailwind's classes. 🧡

But, even better would be if the Cards always look perfectly-sized, depending on their contents, _without_ needing a stepper or Tailwind classes, right? 😇 ("Default correct" is better than "lots of knobs and switches")

File: components/dashboard/src/components/DropDown.tsx
Comment (jankeromnes) [reply components/dashboard/src/components/DropDown.tsx:44]: > It's better to only use left-aligned items inside the dropdown menus. 💡

Ah, sorry, I meant how the collapsible elements horizontally align with respect to the control, e.g.:

<table><tr><td>
Screenshot 2022-05-30 at 16 01 13
</td><td>
Screenshot 2022-05-30 at 16 01 18
</td><td>
Screenshot 2022-05-30 at 16 01 31
</td></tr><tr><td>
Screenshot 2022-05-30 at 16 01 56
</td><td>
Screenshot 2022-05-30 at 16 02 41
</td></tr></table>

**Comment by geropl** [comment]
Starting review now...

**Review by geropl — APPROVED** [review]
@jankeromnes Tested and fixed: Thx for that, and approve! :heavy_check_mark: 

I _think_ I still managed to break it:
image

I think the reasons is the "Quantity" dropdown:
 - during initial upgrade I choose 10 (or sth)
 - I got billed 175 € - without gaining anything :grimacing: 
 - the subsequent 1st quantity upgrade (10 -> 2) did not trigger an addition charge being created (:+1: )
  - the 2nd quantity upgrade (2 -> 3) incurred the charge (+35€) for the current month (:+1: )

IMO we should remove the quantity dropdown if possible.

/hold feel free to address this in this PR or let's discuss/fix in a follow up! :+1:

File: components/dashboard/src/teams/TeamBilling.tsx
Comment (gtsiolis) [reply components/dashboard/src/teams/TeamBilling.tsx:256]: > Also, it's good to keep in mind that Tailwind is so nice because it allows directly "hard-coding" style in a nice way

I'm also in favor of duplication over abstraction as it comes with multiple benefits, but in the long run it makes sense to abstract _some_ component aspects to help a) avoid running into code reviews nitpicking dimensions or other attributes, b) avoid requesting UX reviews when possible to avoid, c) feeling more confident about the product UX, and more.

However, this issue will become more relevant as these components mature and are widely used. For now, as long as we have two component instances of the solid card component, any approach would suffice and does not worth keeping this PR from being merged. 🤝 

> But, even better would be if the Cards always look perfectly-sized, depending on their contents, without needing a stepper or Tailwind classes, right?

Debatable! 🙂 

In most cases, the card component would need to use a specified width or height, which could be challenging to maintain when having multiple cards side-by-side that from a design perspective have to use fixed and specific dimensions regardless if some contents are missing. For example, see editor preferences, etc.

However, there's still need for having flexible card layouts without fixed width that fill the parent container on smaller viewports, etc.

**Comment by jankeromnes** [comment]
> @jankeromnes Tested and fixed: Thx for that, and approve! ✔️

Woohoo 🙌 many thanks!

> - during initial upgrade I choose 10 (or sth)
> - I got billed 175 € - without gaining anything 😬

"User buys too many seats on purpose, is then surprised to pay for too many seats" 😅

All joking aside, I agree that it would be best to hide this quantity selector that isn't needed. It's confusing at best, if not actively harmful.

Will see if there is an easy fix, but if there isn't, I'll just merge this as is and open a follow-up issue.

**Comment by jankeromnes** [comment]
I think the quantity selector can only be disabled in Chargebee's settings:

| Chargebee plan settings | Disallow editing quantity |
| --- | --- |
| Screenshot 2022-05-30 at 18 10 56 | Screenshot 2022-05-30 at 18 10 20 |

Happy to do this, but since it's a follow-up issue, let's unblock this PR first.

/unhold

File: components/dashboard/src/components/DropDown.tsx
Comment (gtsiolis) [reply components/dashboard/src/components/DropDown.tsx:44]: Ah, thanks for clarifying, @jankeromnes!

**Pull request merged**
